fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OutputPropertyDeriver.java

```diff
@@ -0,0 +1,550 @@
+                                                                 List<Integer> rightShuffleColumns) {
+        DistributionSpec.PhysicalPropertyInfo leftInfo = leftScanDistributionSpec.getPhysicalPropertyInfo();
+        DistributionSpec.PhysicalPropertyInfo rightInfo = rightScanDistributionSpec.getPhysicalPropertyInfo();
+
+        ColocateTableIndex colocateIndex = Catalog.getCurrentColocateIndex();
+        long leftTableId = leftInfo.colocateTableList.get(0);
+        long rightTableId = rightInfo.colocateTableList.get(0);
+
+        if (leftTableId == rightTableId && !colocateIndex.isSameGroup(leftTableId, rightTableId)) {
+            return createPropertySetByDistribution(leftScanDistributionSpec);
+        } else {
+            DistributionSpec.PhysicalPropertyInfo newPhysicalPropertyInfo =
+                    new DistributionSpec.PhysicalPropertyInfo();
+            newPhysicalPropertyInfo.addColocateTableList(leftInfo.colocateTableList);
+            newPhysicalPropertyInfo.addColocateTableList(rightInfo.colocateTableList);
+
+            // TODO: compute the shuffle columns, use left and right shuffle columns
+            HashDistributionDesc outputDesc =
+                    new HashDistributionDesc(leftShuffleColumns, HashDistributionDesc.SourceType.SHUFFLE_LOCAL);
+            return createPropertySetByDistribution(new HashDistributionSpec(outputDesc, newPhysicalPropertyInfo));
+        }
+    }
+
+    public boolean canColocateJoin(HashDistributionSpec leftLocalDistributionSpec,
+                                   HashDistributionSpec rightLocalDistributionSpec,
+                                   List<Integer> leftShuffleColumns, List<Integer> rightShuffleColumns) {
+        HashDistributionDesc leftLocalDistributionDesc = leftLocalDistributionSpec.getHashDistributionDesc();
+        HashDistributionDesc rightLocalDistributionDesc = rightLocalDistributionSpec.getHashDistributionDesc();
+
+        if (ConnectContext.get().getSessionVariable().isDisableColocateJoin()) {
+            return false;
+        }
+
+        DistributionSpec.PhysicalPropertyInfo leftInfo = leftLocalDistributionSpec.getPhysicalPropertyInfo();
+        DistributionSpec.PhysicalPropertyInfo rightInfo = rightLocalDistributionSpec.getPhysicalPropertyInfo();
+
+        Preconditions.checkState(!leftInfo.colocateTableList.isEmpty());
+        Preconditions.checkState(!rightInfo.colocateTableList.isEmpty());
+
+        ColocateTableIndex colocateIndex = Catalog.getCurrentColocateIndex();
+        long leftTableId = leftInfo.colocateTableList.get(0);
+        long rightTableId = rightInfo.colocateTableList.get(0);
+
+        // join self
+        if (leftTableId == rightTableId && !colocateIndex.isColocateTable(leftTableId)) {
+            if (!leftInfo.isSinglePartition() || !rightInfo.isSinglePartition() ||
+                    !leftInfo.partitionIds.equals(rightInfo.partitionIds)) {
```

Comment:
check `!leftInfo.partitionIds.equals(rightInfo.partitionIds)` is enough?

Reply:
Any example?